### ISSUE TITLE
Make hypervisor.go and its associated dependencies a self contained package

### ISSUE THREAD

**Issue by mcastelino** [kata-containers/kata-containers#2672]
# Make hypervisor.go and its associated dependencies a self contained package.

Today govmm is a self contained package hosted by Kata containers that can be consumed by other projects.
However govmm does not contain all the logic required to manage a VM. Also that logic is found in virtcontainers.

The goal of making hypervisors its own package us to decouple VM management from virtcontainers (and everything container and Kubernetes specific). This will allow it to be used by other projects. It will also allow us to cleanup some of the logic that has resulted in very tight coupling in virtcontainers where such tight 

**Potential Solution**

In order to minimize disruption to Kata hypervisors can be moved into its own self contained package (and sub packages like devices and types) in multiple phases

## Phase 1:
- Export interfaces and structures and fields to make the VMM logic consumable directly from virtcontainers (along the lines of how vm.go consumes the VM interface but full featured). This will allow the core Kata logic to remain unchanged and will not impact stability of Kata in anyway.
- Add additional methods to the existing hypervisor interface to allow VMs to be managed (this will be in addition to some very container specific methods we have like create sandbox)
- Show that the hypervisor interface can be consumed by a different project 
     - Which is why we need to make the logging and tracing in hypervisors and lower level logic virtcontainers and Kata agnostic
     - This may also mean offering agent free duplicates of some of the methods
- All of Kata's CI should pass without issue. We will need to add additional hypervisor specific tests if needed to cover the new methods

Note: The hypervisor implementation should not assume the existence of the Kata agent or use Kubernetes or containerd specific types.

## Phase 2:
- Rework the sandbox logic to consume the newly added methods

## Phase 3:
- Move the hypervisor logic to its own package (like govmm)

We should ensure that we do not have callbacks from hypervisors into kata-containers (unless it is for self contained utility sub packages like (uuid, signals...)

**Alternatives you've considered**

Move logic from hypervisors to govmm and make govmm generic.

**Comment by fidencio** [comment]
cc @dgibson

**Comment by mcastelino** [comment]
Left unsaid is that if we do this govmm logic will be subsumed into hypervisors and the package will get deprecated.

**Comment by dgibson** [comment]
I do not think this is feasible.  There are a lot of different modes of use for a hypervisor, which makes designing a general purpose abstraction for them an immensely difficult problem.  Note, for example, that this is essentially the entire purpose of libvirt - they have a substantial team who've been working on this for many years, and it's at best a partial abstraction with a lot of issues.

The current hypervisor interface in Kata is built specifically for Kata's needs, and I definitely don't think making it more general is a task the Kata team can reasonably take on.  Indeed, I'm currently working on going the opposite direction: govmm is *also* a poor abstraction and while not explicitly Kata specific, what it does and does not support is largely governed by what Kata has used.  I therefore am intending to replace it with a smaller and unapologetically Kata-specific interface to qemu, which would leave govmm as an orphaned package.

**Comment by dgibson** [comment]
> @dgibson It's a bit different than libvirt. What we want for virtcontainers is a proper abstraction of hypervisor that doesn't necessarily have the idea of sandbox or agent, -- which is the reason for the mess right now. The new interface should focus on VM and its operations. As for libvirt, it should be possible to make it an implementation of the new hypervisor interface, instead of making it kata's hypervisor abstraction. Actually, with the refactoring, it should be pretty easy to do so.

Sure, use of libvirt as a backend is an entirely unrelated topic.

> The refactoring would in effect drop the dependency on govmm but it is not (IMO) the main purpose of the work. The new interface would still be kata-specific and we do not intend to provide it as an external stable API. 

Um.. that doesn't seem to agree with the initial description which says "This will allow it to be used by other projects.".  If it's still an internal interface, what's the benefit to splitting it out into a separate package?

> As proof, runv already has such an interface (https://github.com/hyperhq/runv/blob/master/hypervisor/hypervisor.go) so it is totally doable IMHO.

Well, and we already have `type hypervisor interface`; we certainly can (and must) abstract the hypervisor within the context of Kata.  But the more general we try to make it, the more difficult it gets.

**Comment by bergwolf** [comment]
> Um.. that doesn't seem to agree with the initial description which says "This will allow it to be used by other projects.". If it's still an internal interface, what's the benefit to splitting it out into a separate package?

We can get cleaner interfaces between components inside virtcontainers. We can do it without putting it in a separate package but obviously we've failed on that path. Actually not just hypervisor.go, I would propose we split virtcontainers into several packages and move them to `pkg/` directory, and eventually there is no virtcontainers package.

**Comment by dgibson** [comment]
> > Um.. that doesn't seem to agree with the initial description which says "This will allow it to be used by other projects.". If it's still an internal interface, what's the benefit to splitting it out into a separate package?
> 
> We can get cleaner interfaces between components inside virtcontainers. We can do it without putting it in a separate package but obviously we've failed on that path. Actually not just hypervisor.go, I would propose we split virtcontainers into several packages and move them to `pkg/` directory, and eventually there is no virtcontainers package.

If we haven't managed to clean up the interface in the same package, I don't see much reason to believe we will once it's a separate package - it seems like putting the cart before the horse.  But, whatever, I've no objection on that front.

What does concern me is that the initial description really does seem to imply this will become an externally consumed interface, which I don't think is a good idea.

**Comment by bergwolf** [comment]
@dgibson AFAICT @mcastelino does want to use the hypervisor interafce in other projects. But I don't think we can provide a stable API for him. However, he is still free to use it in different projects even if we might change the APIs from time to time.

**Comment by dgibson** [comment]
@mcastelino any comment here?  As @bergwolf and I have both commented, it's not really within Kata's capacity to provide a stable (even slightly) external interface here.  If you're ok with that, then I have no strong objection, although doing the package split *before* designing the new, improved interface still seems backwards to me.

**Comment by mcastelino** [comment]
> @mcastelino any comment here? As @bergwolf and I have both commented, it's not really within Kata's capacity to provide a stable (even slightly) external interface here. If you're ok with that, then I have no strong objection, although doing the package split _before_ designing the new, improved interface still seems backwards to me.

@dgibson @bergwolf Apologies for the delayed response. I wanted to see how many abstractions leaks we had in the Firecracker and Cloud Hypervisor implementations and there are a few. I will push an update to the PR once I clean them up and document them. I have code now that shows that we can separate the hypervisor logic from Kata.

It should demonstrate that there does exist (and more importantly) should exist a clean boundary between hypervisor logic and Kata specific logic. The code changes needed to use a separate CreateVM (and then use that with other logic in CreateSandbox) is possible. There are some abstractions leaks in the logging and more importantly in the tracing logic that I am trying to figure out a clean way to handle. I will need some help from @jodh-intel here.

IMHO having a clean separation between kata containers and the hypervisor will lead to better more maintainable and testable code overall. And this is something we should have done when we added Firecracker and then Cloud Hypervisor. As to splitting before designing, I feel that a proof of concept that shows a split with the existing code that passes the Kata CI will be better way to start the discussion.

Also there is a history to the hypervisor abstraction. When I added support for Firecracker in Kata we had realized that a lot of the kata code was qemu specific. In the urgency to demonstrate the integration we did not clean it up. When we added support for cloud hypervisor we continued to leverage what existed. Also virtcontainers was created with the goal that it also would be consumed as package cleanly by other projects. That too did not happen. So now we have multiple layers of logic that does not add value and just make changes harder and more complicated than they need to be. 

The networking endpoint is one such example. The fact that we have some device definitions and some are directly within virt containers is yet another example.

Let us try and push the updated code changes this week so that we can have a better discussion and feedback.

I do believe that for certain projects even the limited Hypervisor package that kata can expose would be useful.

**Comment by bergwolf** [comment]
@mcastelino I agree that there should be a clear boundary between sandbox and hypervisor, and that we should introduce such a boundary for kata's own sake. The hypervisor interface may end up being stable, or not. It would be exposed as other packages in kata would consume it too. It is just that we will not be able to make a promise to keep the hypervisor interface stable (nor plan to). So if you use it in other projects, be careful that the interface may change from time to time.

**Comment by dgibson** [comment]
@mcastelino, I'm certainly in favour of cleaner internal interfaces.  It's only the fact that the initial post seemed to be suggesting another maintained external interface that I was concerned.  I can't really judge the new interface until I see the actual patches, but I'll be happy to review.

**Comment by cmaf** [comment]
>  There are some abstractions leaks in the logging and more importantly in the tracing logic that I am trying to figure out a clean way to handle. I will need some help from @jodh-intel here.

@mcastelino What changes do you need to make to the tracing logic?

**Comment by mcastelino** [comment]
> @mcastelino, I'm certainly in favour of cleaner internal interfaces. It's only the fact that the initial post seemed to be suggesting another maintained external interface that I was concerned. I can't really judge the new interface until I see the actual patches, but I'll be happy to review.

@dgibson @bergwolf I have been iterating and trying to identify the gaps and changes needed here.
https://github.com/kata-containers/kata-containers/pull/2713

You will see the besides CreateVM the hypervisor logic is mostly generic. When it comes to createSandbox I have identified what is sandbox specific and what is hypervisor generic.

I also addeda simple test to validate that such an abstraction is even usable.

https://github.com/mcastelino/kata-containers/blob/topic/hypervisor/src/runtime/virtcontainers/hypervisor-abstraction-test/main.go
